[KEP] Maximum Execution Time (#3133)
kubernetes-sigs/kueue: keps/3125-maximum-execution-time/README.md

    - [Unit Tests](#unit-tests)
    - [Integration tests](#integration-tests)
      - [controller/core/workload](#controllercoreworkload)
      - [controller/jobs/job](#controllerjobsjob)
  - [Graduation Criteria](#graduation-criteria)
- [Implementation History](#implementation-history)
- [Drawbacks](#drawbacks)
- [Alternatives](#alternatives)
<!-- /toc -->

## Summary

Add the ability to set a maximum execution time of a Job.

## Motivation

Different Jobs CRDs have a field with similar semantics, but there is no standard. For example `batch/Job` has `spec.activeDeadlneSeconds`, while JobSet does not have such an API for now.

### Goals

We would like to have a common API of setting and enforcing the maximum execution time of a Job.

### Non-Goals

Change the behavior of the Kueue's scheduler based on the maximum execution time.

## Proposal

A new Kueue specific label is added `kueue.x-k8s.io/max-exec-time-seconds` containing
the maximum number of seconds the Job is expected to execute.

The duration is passed to the Job's Workload.

If a workload stays admitted for longer that it is expected it will be Deactivated. 

### User Stories (Optional)

#### Story 1

> **TBD**

Author: done

### Notes/Constraints/Caveats (Optional)

### Risks and Mitigations

> We can accumulate the time a Workload has spent in Admitted state in previous Admit - Evict cycles, however:

Author: > For 1: the deletion of workload could also affect other features such as waitForPodsReady.

Just because other feature are affected it's not an indication that a concern is irrelevant.

> For 2: most AI/ML Jobs would have check-pointing mechanisms to not reset progress on eviction.

Even with a check-pointing mechanism an `Evict - Admit` cycle is adding an overhead and the number of such cycles is very likely unknown to the job submitter.

Given this and the fact that accounting for the previously spent time as Admitted requires additional API changes that will be hard to revert, I  believe it's better to skip the cumulative approach in the initial implementation, and potentially do it in a follow-up.

> We can accumulate the time a Workload has spent in Admitted state in previous Admit - Evict cycles, however:

Reviewer: Not an expert but it seems like what @mimowo is requesting makes sense. @mwielgus WDYT?

> We can accumulate the time a Workload has spent in Admitted state in previous Admit - Evict cycles, however:

Reviewer 2: I agree with @mimowo.

> We can accumulate the time a Workload has spent in Admitted state in previous Admit - Evict cycles, however:

Author: done

- [Alternatives](#alternatives)
<!-- /toc -->

## Summary

Add the ability to set a maximum execution time of a Job.

## Motivation

Extend Kueue so that kjobctl can incorporate functionalities equivalent to `-t` in [slurm](https://slurm.schedmd.com/).

Different Jobs CRDs have a field with similar semantics, but there is no standard. For example `batch/Job` has `spec.activeDeadlneSeconds`, while JobSet does not have such an API for now.

### Goals

Introduce a Job agnostic API of setting and enforcing the maximum execution time of a Job.

### Non-Goals

Change the behavior of the Kueue's scheduler based on the maximum execution time.

## Proposal

A new Kueue specific label is added `kueue.x-k8s.io/max-exec-time-seconds` containing
the maximum number of seconds the Job is expected to execute.

The duration is passed to the Job's Workload.

If a workload stays admitted for longer that it is expected it will be Deactivated. 

### User Stories (Optional)

#### Story 1

As a Batch User, I want to have the ability to set a limit on how long my job can run.

### Notes/Constraints/Caveats (Optional)

### Risks and Mitigations

## Design Details

### API

#### Workload Spec

> Add a new optional filed to hold the 

Reviewer 3: ```suggestion
Add a new optional field to hold the 
```

Author: done

```go
// WorkloadSpec defines the desired state of Workload
// +kubebuilder:validation:XValidation:rule="has(self.priorityClassName) ? has(self.priority) : true", message="priority should not be nil when priorityClassName is set"
type WorkloadSpec struct {
	// ...

	// MaximumExecutionTime if provided, determines the maximum time the workload can be admitted  
	// befrore it's automatically deactivated.
	MaximumExecutionTime *metav1.Duration `json:"maximumExecutionTime,omitempty"`
}
```

#### Workload Status

> Add a optional filed to hold the 

Reviewer 3: ```suggestion
Add a optional field to hold the 
```

Author: done

```go
// WorkloadStatus defines the observed state of Workload
type WorkloadStatus struct {
	// ...

	// AccumulatedExecutionTime holds the total duration the workload spent in Admitted state
	// in the previous `Admit` - `Evict` cycles.

> 	AccumulatedExecutionTime metav1.Duration `json:"maximumExecutionTime,omitempty"`

Reviewer 2: The name of the field suggest that this is the current, often updated value. The description mentions that it is from **previous** cycles. Please consider renaming the field so that it is obvious that the field value is not current and not updated.

Author: done,`AccumulatedPastAdmittedTime`

}

```

### Constants

Define a new constant holding the new label name `kueue.x-k8s.io/max-exec-time-seconds`.

### Controller

#### Workload
During reconcile:

- When a workload transitions out of `Admitted` state, add the time spent in its `status.AccumulatedExecutionTime`
```go
cond := apimeta.FindStatusCondition(wl.Status.Conditions, kueue.WorkloadAdmitted)
wl.Status.AccumulatedExecutionTime += wl.Spec.MaximumExecutionTime.Duration -
    time.Now().Sub(cond.LastTransitionTime.Time)
```
- When the workload is admitted and a maximum execution time is specified compute the remaining execution execution time as:
```go
if wl.Spec.MaximumExecutionTime != nil {
	if cond := apimeta.FindStatusCondition(wl.Status.Conditions, kueue.WorkloadAdmitted); cond != nil && cond.Status == metav1.ConditionTrue {
		remainingTime := wl.Spec.MaximumExecutionTime.Duration -
            wl.Status.AccumulatedExecutionTime -
            time.Now().Sub(cond.LastTransitionTime.Time)
	}
}
```
  - If `remainingTime` > 0 , a reconcile request should be queued with a `remainingTime` delay.

>   - If `remainingTime` <= 0 , the workload should be deactivated `wl.spec.Active = *false`.

Reviewer 2: Will there be an explicit message/event/condition explaining why the workload was deactivated?

Author: sure, and we should also reset the accumulated time, otherwise reactivating the wl will be ineffective

    - [Jobs / Jobframework](#jobs--jobframework)
  - [Webhooks](#webhooks)
    - [Workload](#workload-1)
    - [Jobs](#jobs)
  - [Test Plan](#test-plan)
      - [Prerequisite testing updates](#prerequisite-testing-updates)
    - [Unit Tests](#unit-tests)
    - [Integration tests](#integration-tests)
      - [controller/core/workload](#controllercoreworkload)
      - [controller/jobs/job](#controllerjobsjob)
  - [Graduation Criteria](#graduation-criteria)
- [Implementation History](#implementation-history)
- [Drawbacks](#drawbacks)
- [Alternatives](#alternatives)
<!-- /toc -->

## Summary

Add the ability to set a maximum execution time of a Job.

## Motivation

Extend Kueue so that kjobctl can incorporate functionalities equivalent to `-t` in [slurm](https://slurm.schedmd.com/).

Different Jobs CRDs have a field with similar semantics, but there is no standard. For example `batch/Job` has `spec.activeDeadlneSeconds`, while JobSet does not have such an API for now.

### Goals

Introduce a Job agnostic API of setting and enforcing the maximum execution time of a Job.

### Non-Goals

- Change the behavior of the Kueue's scheduler based on the maximum execution time.

> - Mitigate potential conflicting configurations between `kueue.x-k8s.io/max-exec-time-seconds` and other job specific `activeDeadline` functionalities.

Reviewer 4: I guess that we can expand the `GenericJob` interface or introduce new interface, then we can add the validation to verify if Job does not have `activeDeadlineSeconds` and `kueue.x-k8s.io/max-exec-time-seconds` annotations.

But, users may want to specify both. So, I will postpone this decision to GA criteria. So, could you add an evaluation to to GA criteria if we should extend the JobFramework interface solution to support validations between `kueue.x-k8s.io/max-exec-time-seconds` and other job specific `activeDeadline` functionalities?

Reviewer 4: Oh, sorry. That's true.
I wanted to say the validation implementation in the https://github.com/kubernetes-sigs/kueue/blob/main/pkg/controller/jobs/job/job_webhook.go.

However, you do not plan to implement such validations for now, right?

Author: right

## Proposal

A new Kueue specific label is added `kueue.x-k8s.io/max-exec-time-seconds` containing
the maximum number of seconds the Job is expected to execute.

The duration is passed to the Job's Workload.

If a workload stays admitted for longer that it is expected it will be Deactivated. 

### User Stories (Optional)

#### Story 1

As a Batch User, I want to have the ability to set a limit on how long my job can run.

### Notes/Constraints/Caveats (Optional)

### Risks and Mitigations

## Design Details

### API

#### Workload Spec

Add a new optional field to hold the 

```go
// WorkloadSpec defines the desired state of Workload
// +kubebuilder:validation:XValidation:rule="has(self.priorityClassName) ? has(self.priority) : true", message="priority should not be nil when priorityClassName is set"
type WorkloadSpec struct {
	// ...

	// MaximumExecutionTime if provided, determines the maximum time the workload can be admitted  
	// befrore it's automatically deactivated.

> 	MaximumExecutionTime *metav1.Duration `json:"maximumExecutionTime,omitempty"`

Reviewer 5: ```suggestion
	MaximumExecutionTimeSeconds int32 `json:"maximumExecutionTimeSeconds,omitempty"`
```
https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#units

Reviewer 5: or int64

}
```

#### Workload Status

Add a optional field to hold the 
```go
// WorkloadStatus defines the observed state of Workload
type WorkloadStatus struct {
	// ...

	// AccumulatedPastAdmittedTime holds the total duration the workload spent in Admitted state
	// in the previous `Admit` - `Evict` cycles.

> 	AccumulatedPastAdmittedTime *metav1.Duration `json:"AccumulatedPastAdmittedTime,omitempty"`

Reviewer 5: also int32

Reviewer 5: The name of this field doesn't have a direct correlation to `maximumExecutionTime`. So this should either be `accumulatedPastExecutionTime` or the other field should be `maximumAdmittedTime`

}

```

### Constants

Define a new constant holding the new label name `kueue.x-k8s.io/max-exec-time-seconds`.

### Controller

#### Workload
During reconcile:

- When a workload transitions out of `Admitted` state, add the time spent in its `status.AccumulatedPastExecutionTimeSeconds`
```go
cond := apimeta.FindStatusCondition(wl.Status.Conditions, kueue.WorkloadAdmitted)
wl.Status.AccumulatedPastExecutionTimeSeconds += wl.Spec.MaximumExecutionTimeSeconds -
    time.Now().Sub(cond.LastTransitionTime.Time).Seconds()
```
- When the workload is admitted and a maximum execution time is specified compute the remaining execution execution time as:
```go
if wl.Spec.MaximumExecutionTimeSeconds != nil {
	if cond := apimeta.FindStatusCondition(wl.Status.Conditions, kueue.WorkloadAdmitted); cond != nil && cond.Status == metav1.ConditionTrue {
		remainingTime := wl.Spec.MaximumExecutionTimeSeconds -
            wl.Status.AccumulatedPastExecutionTimeSeconds -
            time.Now().Sub(cond.LastTransitionTime.Time).Seconds()
	}
}
```
  - If `remainingTime` > 0 , a reconcile request should be queued with a `remainingTime` delay.

>   - If `remainingTime` <= 0 , the workload should be deactivated `wl.spec.Active = *false`, the `wl.status,AccumulatedPastExecutionTimeSeconds` set to 0, and a relevant event recorded.

Reviewer 4: Shouldn't we follow the current deactivation mechanism instead of directly deactivating it?
`DeactivationTarget` -> `.spec.active=false`. I would like to keep consistency between all deactivation mechanisms, avoiding struggling.

Reviewer 6: Doing this is the follow up reconcile is not a problem. 

I also want to use the same mechanism via DeactivationTarget. 

The intention of introducing DeactivationTarget was to also use it for ProvReq retries, so if it wasn't done I consider it an omission. 
cc @PBundyra who is currently working on the changes to the retry mechanism.

Reviewer 4: For AdmissionCheck, I was supposed to switch the DeactivationTarget mechanism by @PBundyra at https://github.com/kubernetes-sigs/kueue/issues/1353.

Author: done

>   - If `remainingTime` <= 0 , the workload should be deactivated `wl.spec.Active = *false`, the `wl.status,AccumulatedPastExecutionTimeSeconds` set to 0, and a relevant event recorded.

Reviewer 4: ```suggestion
  - If `remainingTime` <= 0 , the workload should be deactivated `wl.spec.Active = *false`, the `wl.status.accumulatedPastExecutionTimeSeconds` set to 0, and a relevant event recorded.
```

Author: done

> 

Reviewer 4: What if the Workload is preempted / evicted even if the remainingTime is greater than 0?
Will we reset the `.status.accumulatedPastExecutionTimeSeconds`?

Author: The time spent as admitted in the current cycle is added to `accumulatedPastExecutionTimeSeconds`. L120

Reviewer 4: I see. So, when we re-admit the Workload after preemption, we will add the `.spec. maximumExecutionTimeSeconds ` to the `.status. accumulatedPastExecutionTimeSeconds `, right?

Author: No, we are accumulating the time the workload was in admitted state , regardless of how many admit/evict cycles the wl has if the sum gets to  `maximumExecutionTimeSeconds` we are deactivating. 

(The accumulated time resets on deactivation).

Reviewer 4: Oh, I see. In that case, specifying both Job activeDeadline and this maximumExecutionTimeSeconds could cooperate since we can specify the activeDeadline per oneshot, but we can specify the maximumExecutionTime based on cluster-level Job running limitation time.

#### Jobs / Jobframework

> Upon workload creation, the value from `kueue.x-k8s.io/max-exec-time-seconds` is converted to `metav1.Duration` and set into the new workload's `spec`.

Author: done

### Webhooks

#### Workload

> Validate that if specified the value of `wl.spec.maximumExecutionTime` is grater then 0.

Reviewer 5: ```suggestion
Validate that if specified the value of `wl.spec.maximumExecutionTime` is greater than 0.
```

Optionally: `wl.spec.maximumExecutionTime` is immutable [ while the workload is admitted ]

#### Jobs

> Validate that the value provided in the `kueue.x-k8s.io/max-exec-time-seconds` label is the base 10 representation of an integer grater then 0.

Reviewer 5: ```suggestion
Validate that the value provided in the `kueue.x-k8s.io/max-exec-time-seconds` label is the base 10 representation of an integer greater than 0.
```

Author: done

to implement this enhancement.

##### Prerequisite testing updates

No regressions in the existing tests.

#### Unit Tests

As needed to provide coverage fore the new code.

-  pkg/controller/core: 2024-09-25 - 9.9%
-  pkg/controller/jobframework: 2024-09-25 - 1.9%
-  pkg/controller/jobs/deployment: 2024-09-25 - 0.4%
-  pkg/controller/jobs/job: 2024-09-25 - 8.5%
-  pkg/controller/jobs/jobset: 2024-09-25 - 4.2%
-  pkg/controller/jobs/kubeflow/jobs/mxjob: 2024-09-25 - 3.6%
-  pkg/controller/jobs/kubeflow/jobs/paddlejob: 2024-09-25 - 3.5%
-  pkg/controller/jobs/kubeflow/jobs/pytorchjob: 2024-09-25 - 0.7%
-  pkg/controller/jobs/kubeflow/jobs/tfjob: 2024-09-25 - 0.7%
-  pkg/controller/jobs/kubeflow/jobs/xgboostjob: 2024-09-25 - 3.5%
-  pkg/controller/jobs/kubeflow/kubeflowjob: 2024-09-25 - 0.0%
-  pkg/controller/jobs/mpijob: 2024-09-25 - 2.2%
-  pkg/controller/jobs/pod: 2024-09-25 - 8.9%
-  pkg/controller/jobs/raycluster: 2024-09-25 - 3.8%
-  pkg/controller/jobs/rayjob: 2024-09-25 - 1.2%
-  pkg/webhooks: 2024-09-25 - 2.1%

#### Integration tests

##### controller/core/workload

Add "An admitted workload is deactivated when its maximum execution time expires"

##### controller/jobs/job

> Add "An job is suspended when its maximum execution time expires"

Reviewer 5: ```suggestion
Add "A job is suspended when its maximum execution time expires"
```

## Motivation

Extend Kueue so that kjobctl can incorporate functionalities equivalent to `-t` in [slurm](https://slurm.schedmd.com/).

Different Jobs CRDs have a field with similar semantics, but there is no standard. For example `batch/Job` has `spec.activeDeadlneSeconds`, while JobSet does not have such an API for now.

### Goals

Introduce a Job agnostic API of setting and enforcing the maximum execution time of a Job.

### Non-Goals

- Change the behavior of the Kueue's scheduler based on the maximum execution time.
- Mitigate potential conflicting configurations between `kueue.x-k8s.io/max-exec-time-seconds` and other job specific `activeDeadline` functionalities.

## Proposal

A new Kueue specific label is added `kueue.x-k8s.io/max-exec-time-seconds` containing
the maximum number of seconds the Job is expected to execute.

The duration is passed to the Job's Workload.

If a workload stays admitted for longer that it is expected it will be Deactivated. 

### User Stories (Optional)

#### Story 1

As a Batch User, I want to have the ability to set a limit on how long my job can run.

### Notes/Constraints/Caveats (Optional)

### Risks and Mitigations

> ## Design Details

Reviewer 6: Please make the label immutable.

The reason is that when we use it for slurm-like scheduling the one could abuse the label by setting very small value on admission and extending after admission. 

We could mitigate this by evicting on changes, but it adds complexity so it might be better to revisit mutability for GA. 

cc @tenzen-y @alculquicondor
